lib/portlayer/exec/commit_test.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+func TestBatchBlockOnFunc(t *testing.T) {
+	ctx := context.Background()
+
+	operation := func(op trace.Operation) error {
+		time.Sleep(10 * time.Millisecond)
+		return nil
+	}
+
+	// test fetch one request
+	testFetchOne(t, ctx, operation)
+
+	// serialize test: batch size 1, 3 requests
+	// batch size 1 = no batching. Process requests one at a time
+	// total time spent should be within: (# of requests) * (operation wait time)
+	testMultipleBatch(t, ctx, 1, 3, operation)
+
+	// batch size 10, 5 requests
+	// the 5 requests should be processed in one batch
+	// total time spent should be within: 1 * (operation wait time)
+	testMultipleBatch(t, ctx, 10, 5, operation)
+
+	//  batch size 100, 50 requests
+	// the 50 requests should be processed in one batch
+	// total time spent should be within: 1 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 50, operation)
+
+	// stress test: batch size 100, 200 requests
+	// the 200 requests should be processed in 2 batches
+	// total time spent should be within: 2 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 200, operation)
+
+	// stress test: batch size 100, 500 requests
+	// the 500 requests should be processed in 5 batches
+	// total time spent should be within: 5 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 500, operation)
+}
+
+func testFetchOne(t *testing.T, ctx context.Context, operation func(op trace.Operation) error) {
+	batch := make(chan chan error, 5) // batch size 5
+
+	// fire background reader
+	go batchBlockOnFunc(ctx, batch, operation)
+
+	// send only 1 request
+	sendRequest(t, batch, nil)
+	close(batch)
+}
+
```

Comment:
A test to ensure that `err` propagates back to `sendRequest` properly in error cases